staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -265,38 +293,25 @@ func (r *crdHandler) removeDeadStorage() {
 		if !found {
 			glog.V(4).Infof("Removing dead CRD storage for %v", s.requestScope.Resource)
 			s.storage.DestroyFunc()
-			delete(storageMap, uid)
+			delete(storageMap2, uid)
 		}
 	}
-
-	r.customStorageLock.Lock()
-	defer r.customStorageLock.Unlock()
-
 	r.customStorage.Store(storageMap)
 }
 
 // GetCustomResourceListerCollectionDeleter returns the ListerCollectionDeleter for
 // the given uid, or nil if one does not exist.
 func (r *crdHandler) GetCustomResourceListerCollectionDeleter(crd *apiextensions.CustomResourceDefinition) finalizer.ListerCollectionDeleter {
-	info, err := r.getServingInfoFor(crd)
+	info, err := r.getOrCreateServingInfoFor(crd)
 	if err != nil {
 		utilruntime.HandleError(err)
 	}
 	return info.storage
 }
 
-func (r *crdHandler) getServingInfoFor(crd *apiextensions.CustomResourceDefinition) (*crdInfo, error) {
+func (r *crdHandler) getOrCreateServingInfoFor(crd *apiextensions.CustomResourceDefinition) (*crdInfo, error) {
 	storageMap := r.customStorage.Load().(crdStorageMap)
-	ret, ok := storageMap[crd.UID]
-	if ok {
-		return ret, nil
-	}
-
-	r.customStorageLock.Lock()
-	defer r.customStorageLock.Unlock()
-
-	ret, ok = storageMap[crd.UID]
```

Comment:
I just deleted this duplicated .

Reply:
It was intentional, and was more correct originally (was missing a load after the lock). With your current changes, two threads can both build and set CRD storage.  Restore this section, with the addition of calling Load a second time after locking for write.

The flow should be:

1. Lockless check (load, check, return if found)
2. Lock for write 
3. Recheck (load, check, return if found)
4. Build, clone, set, store

Reply:
Really admire this way to improve performance.